pkg/error_code/error_code.go

```diff
@@ -322,12 +339,105 @@ type notFoundErr struct{ CodedError }
+// A given error code can be triggered by multiple different operations,
+// just as a given operation could result in multiple different error codes.
+//
+// GetOperation is defined, but generally the operation should be retrieved with Operation().
+// Operation() will check if a HasOperation interface exists.
+// As an alternative to defining this interface
+// you can use an existing wrapper (OpErrCode via AddOp) or embedding (EmbedOp) that has already defined it.
+type HasOperation interface {
+	GetOperation() string
+}
+
+// Operation will return an operation string if it exists.
+// It checks for the HasOperation interface.
+// Otherwise it will return the zero value (empty) string.
+func Operation(v interface{}) string {
+	var operation string
+	if hasOp, ok := v.(HasOperation); ok {
+		operation = hasOp.GetOperation()
+	}
+	return operation
+}
+
+// EmbedOp is designed to be embedded into your existing error structs.
+// It provides the HasOperation interface already, which can reduce your boilerplate.
+type EmbedOp struct{ Op string }
+
+// GetOperation satisfies the HasOperation interface
+func (e EmbedOp) GetOperation() string {
+	return e.Op
+}
+
+// MakeOp is just an alternative way to construct EmbedOp.
+func MakeOp(operation string) EmbedOp { return EmbedOp{Op: operation} }
+
+// OpErrCode is an ErrorCode with an "Operation" field attached.
+// This may be used as a convenience to record the operation information for the error.
+// However, it isn't required to be used, see the HasOperation documentation for alternatives.
+type OpErrCode struct {
+	Operation string
+	Err       ErrorCode
+}
+
+// Error prefixes the operation to the underlying Err Error.
+func (e OpErrCode) Error() string {
+	return e.Operation + ": " + e.Err.Error()
+}
+
+// GetOperation satisfies the HasOperation interface.
+func (e OpErrCode) GetOperation() string {
+	return e.Operation
+}
+
+// Code returns the unerlying Code of Err.
+func (e OpErrCode) Code() Code {
+	return e.Err.Code()
+}
+
+// GetClientData returns the ClientData of the underlying Err.
+func (e OpErrCode) GetClientData() interface{} {
+	return ClientData(e.Err)
+}
+
+var _ ErrorCode = (*OpErrCode)(nil)     // assert implements interface
+var _ HasClientData = (*OpErrCode)(nil) // assert implements interface
+var _ HasOperation = (*OpErrCode)(nil)  // assert implements interface
+
+// A Modifier returns a new modified ErrorCode
+type Modifier func(ErrorCode) ErrorCode
+
+// Modify applies the Modifier to an ErrorCode
+func (modifier Modifier) Modify(err ErrorCode) ErrorCode {
+	return modifier(err)
+}
+
+/*
+Op adds an operation to an ErrorCode with Modify.
+This converts the error to the type OpErrCode.
+
+op := errcode.Op("path.move.x")
+if start < obstable && obstacle < end  {
+	return op.Modify(PathBlocked{start, end, obstacle})
+}
+*/
+func Op(operation string) Modifier {
+	return func(err ErrorCode) ErrorCode {
+		return OpErrCode{Operation: operation, Err: err}
+	}
+}
```

Comment:
IMO, the `Modifier` and `Op()` are overly abstracted. How about simply use a `SetOp` function?
```
op := "path.move.x"
if start < obstable && obstacle < end  {
        return errcode.SetOp(PathBlocked{start, end, obstacle}, op)
}
```

Reply:
In [previous commits](https://github.com/pingcap/pd/pull/1167/commits/4e2fb1a3bb7c871657d90b56bceb38b1c46546a4#diff-d412b9240514b30a308e4b4c8208ae72R354) I did have only that function (called `AddOp`).

The `Modifier` abstraction is not that useful as an abstraction because there aren't many other things one would want to modify an error with. Instead the goal is really just to use method chaining instead of package-qualified functions.

Here is the [usage site](https://github.com/pingcap/pd/pull/1167/files#diff-d412b9240514b30a308e4b4c8208ae72R346) for this. I think the extra work to wrap the op variable with `errcode.Op` is actually helpful to understand that it is an operation as understood by errcode. I was thinking of changing the name `Modify` to something like `AddTo`. Then the comparison would be:

```
op.AddTo(core.StoreTombstonedErr{StoreID: storeID})
errcode.SetOp(op, core.StoreTombstonedErr{StoreID: storeID})
```

So you can see I wasn't sure about the best approach here (I also tried out `MakeOp`).
Let me know if you still like the `SetOp` approach the best.

Reply:
em, can we do something like this?
```
errcode.New(PathBlocked{start, end, obstacle}).WithOp("path.move.x")
```
Use `errcode.New` to wrap user defined error code to a type that we can add some chaining methods to.

Reply:
Yes, the use of an interface prevents chaining, and it would be nice to chain to the right. I think though that the `errcode.New` suggestion is not as good as the `errcode.AddOp` function. It requires two method calls to achieve its purpose. In general error codes will look like the [usage site for this PR](https://github.com/pingcap/pd/pull/1167/files#diff-d412b9240514b30a308e4b4c8208ae72R346) where there is one operation in a function and multiple possible error codes: in this case the `Modify` form is doing just one method call and the additional wrapping is amortized over usage in the function.

Reply:
I just added a commit that maintains the `Modify` style but switches to concrete types and names. Let me know if you think this is any better or if you still prefer `errcode.AddOp(string, ErrorCode)`

https://github.com/pingcap/pd/pull/1167/commits/bd0a49595021ee16439d721c9772b0651e342c37

Reply:
Do we need to add more properties like `Operation` to `ErrorCode`s in the future? If so, the style I suggested would be better: `errcode.New(UserType).WithOp("user.op").WithFoo("bar")` compare to `foo.AddTo("bar", op.AddTo("op", UserType))`.

Reply:
It is also possible to compose the modifiers before applying them. `foo.And(op).AddTo()`.
I don't foresee other properties that will be used frequently like this. Only some that would be added during an error constructor or in a single spot before errors are sent back. In upspin they added the user, but it seems to me that the user should be added when errors are sent back and the sender already knows who they are. They also added the file, which is a common resource for them, but I think our resources are more diverse and will be part of the error type.